pkg/controller/dnsrecord/actuator.go

```diff
@@ -0,0 +1,142 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dnsrecord
+
+import (
+	"context"
+	"fmt"
+
+	extensionscontroller "github.com/gardener/gardener/extensions/pkg/controller"
+	"github.com/gardener/gardener/extensions/pkg/controller/dnsrecord"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	extensionsv1alpha1helper "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1/helper"
+	"github.com/go-logr/logr"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/gardener/gardener-extension-provider-aws/pkg/aws"
+	awsclient "github.com/gardener/gardener-extension-provider-aws/pkg/aws/client"
+)
+
+type actuator struct {
+	client client.Client
+	logger logr.Logger
+}
+
+func NewActuator(logger logr.Logger) dnsrecord.Actuator {
+	return &actuator{
+		logger: logger.WithName("aws-dnsrecord-actuator"),
+	}
+}
+
+func (a *actuator) InjectClient(client client.Client) error {
+	a.client = client
+	return nil
+}
+
+// Reconcile reconciles the DNSRecord.
+func (a *actuator) Reconcile(ctx context.Context, dns *extensionsv1alpha1.DNSRecord, cluster *extensionscontroller.Cluster) error {
+	// Create AWS client
+	awsClient, err := aws.NewClientFromDNSSecretRef(ctx, a.client, dns.Spec.SecretRef, dns.Spec.Region)
+	if err != nil {
+		return fmt.Errorf("could not create AWS client: %+v", err)
+	}
+
+	// Determine DNS hosted zone ID
+	zone, err := a.getZone(ctx, dns, awsClient)
+	if err != nil {
+		return err
+	}
+
+	// Create or update DNS record
+	ttl := extensionsv1alpha1helper.GetDNSRecordTTL(dns.Spec.TTL)
+	a.logger.Info("Creating or updating DNS record", "zone", zone, "name", dns.Spec.Name, "type", dns.Spec.RecordType, "values", dns.Spec.Values)
+	if err := awsClient.CreateOrUpdateDNSRecord(ctx, zone, dns.Spec.Name, string(dns.Spec.RecordType), dns.Spec.Values, ttl); err != nil {
+		return fmt.Errorf("could not create or update DNS record in zone %s with name %s, type %s, and values %v: %+v", zone, dns.Spec.Name, dns.Spec.RecordType, dns.Spec.Values, err)
+	}
+
+	// Update resource status
+	return extensionscontroller.TryUpdateStatus(ctx, retry.DefaultBackoff, a.client, dns, func() error {
+		dns.Status.Zone = &zone
+		return nil
+	})
+}
+
+// Delete deletes the DNSRecord.
+func (a *actuator) Delete(ctx context.Context, dns *extensionsv1alpha1.DNSRecord, cluster *extensionscontroller.Cluster) error {
```

Comment:
Ok, this might be a bit theoretical and I'm not sure, if we need to think about this case, but hear me out :)

If the record type changes for some reason (might be because the corresponding LB changes from IP to Hostname, or because the order in the values list is changed and now the first one isn't part of a canonical zone anymore, so we change from ALIAS to CNAME, or whatever else), we will orphan the old record in the infrastructure.

If this is indeed a problem, we might want to track the record type that was created in the infrastructure and accordingly a) deny switching the record type or b) delete the old entry before creating one of the new type, or something similar.

Actually, the same applies to a changing name field value (though this one should definitely not change, I suppose). So another option might be to make those fields immutable, though I'm not sure if that's easily achievable via CRDs or if we would need validating webhooks.

Reply:
I don't think this is what will happen if the record type changes. If you specify a record with the correct name and wrong record type, you will get an error from route53 different from "not found". So the record will not be orphaned, instead the deletion will fail. Perhaps this is good enough for a "theoretical" issue like this?

Regarding the name, this should definitely not change. It should be made immutable (and perhaps the record type as well). I will check what is the best way to do it and add it to https://github.com/gardener/gardener/pull/4231.

Reply:
> It seems the behavior depends on the record type ...

In fact I was wrong, the same happens when trying to switch from A to CNAME and vice versa - the update failed as I expected, but then the deletion of the DNSRecord succeeded (it failed on route53 but the error returned was "not found" and so ignored), and so the actual record in route53 was orphaned.

Unresolving this since it's not so trivial, we should at least give it some thought.

Reply:
We discussed this issue and decided not to introduce more complexity at this stage, apart from making the `recordType` field also immutable and making sure that it's actually validated by Gardener.